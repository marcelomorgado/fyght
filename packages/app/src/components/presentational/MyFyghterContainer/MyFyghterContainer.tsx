import React, { useEffect, useState } from "react";
import { Row, Col, Divider } from "antd";
import { MyFyghter } from "../MyFyghter";
import { useFyghtState } from "../../../state";
import { FyghterCreationModal } from "../FyghterCreationModal";

export const MyFyghterContainer: React.FC = () => {
  const [isLoading, setLoading] = useState(true);

  const [
    {
      myFyghter,
      metamask: { account },
    },
    { fetchMyFyghter },
  ] = useFyghtState();

  useEffect(() => {
    const init = async (): Promise<void> => {
      fetchMyFyghter();
      setLoading(false);
    };
    init();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [account]);

  const hasFyghter = !isLoading && myFyghter !== null;

  return (
    <>
      <Divider orientation="left" style={{ color: "#333", fontWeight: "normal" }}>
        My Fyghter
      </Divider>
      <Row gutter={[16, 24]}>
        <Col span={24}>
          {isLoading ? <>{`Loading...`}</> : !hasFyghter ? <FyghterCreationModal /> : <MyFyghter fyghter={myFyghter} />}
        </Col>
      </Row>
    </>
  );
};
