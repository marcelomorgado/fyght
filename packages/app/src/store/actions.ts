import {
  ethers,
  ContractTransaction,
  Event,
  Transaction,
  ContractReceipt,
  Contract,
} from "ethers";
import { BigNumber } from "ethers";

const { getAddress } = ethers.utils;

export const RENAME = "RENAME";
export const CHANGE_SKIN = "CHANGE_SKIN";
export const INCREMENT_MY_FIGHTER_XP = "INCREMENT_MY_FIGHTER_XP";
export const INCREMENT_ENEMY_XP = "INCREMENT_ENEMY_XP";
export const LOAD_ENEMIES = "LOAD_ENEMIES";
export const SET_MY_FYGHTER = "SET_MY_FYGHTER";
export const UPDATE_METAMASK_ACCOUNT = "UPDATE_METAMASK_ACCOUNT";
export const UPDATE_METAMASK_NETWORK = "UPDATE_METAMASK_NETWORK";
export const INITIALIZE_METAMASK = "INITIALIZE_METAMASK";

interface NewFyghter {
  owner: string;
  id: BigNumber;
  name: string;
}

export const createActions = (dispatch: any, state: FyghtContext): any => {
  const incrementMyFyghterXp = (): void =>
    dispatch({ type: INCREMENT_MY_FIGHTER_XP, payload: {} });

  const incrementEnemyXp = (enemyId: BigNumber): void =>
    dispatch({ type: INCREMENT_ENEMY_XP, payload: { enemyId } });

  const setEnemies = (enemies: Fyghter[]): void =>
    dispatch({ type: LOAD_ENEMIES, payload: { enemies } });

  const setMyFyghter = (myFyghter: Fyghter): void =>
    dispatch({ type: SET_MY_FYGHTER, payload: { myFyghter } });

  const createFyghter = async (name: string): Promise<void> => {
    const {
      metamask: { contract: fyghters },
    } = state;

    try {
      const tx: ContractTransaction = await fyghters.create(name);
      await tx.wait();
      console.log(tx);

      // TODO: Get event from transaction
      const filter = fyghters.filters.NewFyghter(null, null, null);
      fyghters.on(filter, async (owner: string, id: number, name: string) => {
        const myFyghter = await fyghters.fyghters(id);
        setMyFyghter(myFyghter);
      });
    } catch (e) {
      console.log(e);
      // Revert message
      //console.log(e.data.message);
    }
  };

  const optimisticUpdate = async (
    tx: Transaction,
    onSuccess: () => void,
    onError: () => void
  ): Promise<void> => {
    const {
      metamask: { provider },
    } = state;

    // Optimistic update
    onSuccess();

    provider.once(tx.hash, (receipt: ContractReceipt) => {
      const { status } = receipt;
      if (!status) {
        onError();
      }
    });
  };

  const renameMyFyghter = async (newName: string): void => {
    const {
      myFyghter,
      metamask: { contract: fyghters },
    } = state;
    const { id: myFyghterId, name: oldName } = myFyghter;

    const tx: ContractTransaction = await fyghters.rename(myFyghterId, newName);

    optimisticUpdate(
      tx,
      () => {
        dispatch({ type: RENAME, payload: { name: newName } });
      },
      () => {
        dispatch({ type: RENAME, payload: { name: oldName } });
      }
    );
  };

  const changeMyFyghterSkin = async (newSkin: string): Promise<void> => {
    const {
      myFyghter,
      metamask: { contract: fyghters },
    } = state;
    const { id: myFyghterId, skin: oldSkin } = myFyghter;

    const tx: ContractTransaction = await fyghters.changeSkin(
      myFyghterId,
      newSkin
    );

    optimisticUpdate(
      tx,
      () => {
        dispatch({ type: CHANGE_SKIN, payload: { skin: newSkin } });
      },
      () => {
        dispatch({ type: RENAME, payload: { name: oldSkin } });
      }
    );
  };

  const attackAnEnemy = async (enemyId: BigNumber): Promise<void> => {
    const {
      myFyghter: { id: myFyghterId },
      metamask: { contract: fyghters, provider },
    } = state;

    const tx: ContractTransaction = await fyghters.attack(myFyghterId, enemyId);

    provider.once(tx.hash, ({ status }: ContractReceipt) => {
      if (!status) {
        // TODO: Error message
      }
    });

    const filter = fyghters.filters.Attack(myFyghterId, null, null);
    fyghters.once(
      filter,
      async (
        myFighterId: BigNumber,
        enemyId: BigNumber,
        winnerId: BigNumber
      ) => {
        if (winnerId.eq(myFighterId)) {
          incrementMyFyghterXp();
        } else {
          incrementEnemyXp(enemyId);
        }
      }
    );
  };

  //
  // Note: This low level code needed to get past events is the way that ethers.js v4 works
  // See more: https://github.com/marcelomorgado/fyght/issues/78
  //
  const loadEnemies = async (): Promise<void> => {
    const {
      metamask: { contract: fyghters, provider, account },
    } = state;

    if (!fyghters || !provider) {
      setEnemies([]);
      return;
    }

    const filter = fyghters.filters.NewFyghter(null, null, null);
    const logs = await fyghters.queryFilter(filter, 0, "latest");
    const enemiesIds = logs
      .map((l: Event) => l.args)
      .filter(
        ({ owner }: NewFyghter) => getAddress(owner) !== getAddress(account)
      )
      .map(({ id }: NewFyghter) => id);

    const enemiesPromises = enemiesIds.map((id: BigNumber) =>
      fyghters.fyghters(id)
    );
    const enemies: Fyghter[] = await Promise.all(enemiesPromises);

    setEnemies(enemies);
  };

  const loadMyFyghter = async (): Promise<void> => {
    const {
      metamask: { contract: fyghters, account },
    } = state;

    const filter = fyghters.filters.NewFyghter(getAddress(account), null, null);
    const logs = await fyghters.queryFilter(filter, 0, "latest");
    const [myFyghterId] = logs
      .map((l: Event) => l.args)
      .map(({ id }: NewFyghter) => id);

    if (myFyghterId) {
      const myFyghter = await fyghters.fyghters(myFyghterId);
      setMyFyghter(myFyghter);
    }
  };

  const setMetamaskAccount = (account: string): void =>
    dispatch({ type: UPDATE_METAMASK_ACCOUNT, payload: { account } });

  const setMetamaskNetworkId = (networkId: number): void =>
    dispatch({ type: UPDATE_METAMASK_NETWORK, payload: { networkId } });

  const initializeMetamask = (): void => {
    const {
      metamask: { ethereum },
    } = state;

    if (ethereum) {
      // Note: The metamask docs recommends to use the 'chainChanged' event instead but it isn't working
      // See more: https://docs.metamask.io/guide/ethereum-provider.html#methods-new-api
      ethereum.on("networkChanged", (networkId: number) => {
        setMetamaskNetworkId(networkId);
      });

      ethereum.on("accountsChanged", ([account]: string[]) => {
        setMetamaskAccount(account);
      });
    }

    dispatch({ type: INITIALIZE_METAMASK, payload: {} });
  };

  return {
    renameMyFyghter,
    changeMyFyghterSkin,
    attackAnEnemy,
    loadEnemies,
    loadMyFyghter,
    setMetamaskAccount,
    setMetamaskNetworkId,
    initializeMetamask,
    createFyghter,
  };
};
