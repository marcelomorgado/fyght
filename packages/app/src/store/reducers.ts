import { ethers } from "ethers";
import { FyghtersFactory } from "../contracts/FyghtersFactory";
import { Fyghters } from "../contracts/Fyghters";
import {
  RENAME,
  CHANGE_SKIN,
  UPDATE_ENEMY_XP,
  UPDATE_MY_FIGHTER_XP,
  LOAD_ENEMIES,
  SET_MY_FYGHTER,
  UPDATE_METAMASK_ACCOUNT,
  UPDATE_METAMASK_NETWORK,
  INITIALIZE_METAMASK,
} from "./actions";
import { Metamask, Fyghter, Action, FyghtContextInterface } from "../global";

declare global {
  interface Window {
    // TODO: Set properly type
    ethereum: any;
  }
}

const { ethereum } = window;
ethereum.autoRefreshOnNetworkChange = false;

export const initialState: FyghtContextInterface = {
  myFyghter: null,
  enemies: [],
  metamask: {
    networkId: ethereum.networkVersion,
    account: null,
    ethereum,
    contract: null,
    provider: null,
  },
};

const myFyghterReducer = (
  state: Fyghter = initialState.myFyghter,
  action: Action
): Fyghter => {
  const { type, payload } = action;
  const { name, skin, xp, myFyghter } = payload;

  switch (type) {
    case RENAME:
      return { ...state, name };
    case CHANGE_SKIN:
      return { ...state, skin };
    case UPDATE_MY_FIGHTER_XP:
      return { ...state, xp };
    case SET_MY_FYGHTER:
      return myFyghter;
    default:
      return state;
  }
};

const enemiesReducer = (
  state: Array<Fyghter> = initialState.enemies,
  action: Action
): Array<Fyghter> => {
  const { type, payload } = action;
  const { enemyId, xp, enemies } = payload;

  switch (type) {
    case UPDATE_ENEMY_XP:
      return state.map(e => {
        if (e.id === enemyId) {
          return { ...e, xp };
        } else {
          return e;
        }
      });
    case LOAD_ENEMIES:
      return enemies;
    default:
      return state;
  }
};

const metamaskReducer = (
  state: Metamask = initialState.metamask,
  action: Action
): Metamask => {
  const { type, payload } = action;
  const { networkId, account } = payload;

  switch (type) {
    case UPDATE_METAMASK_ACCOUNT:
      return { ...state, account };
    case UPDATE_METAMASK_NETWORK:
      return { ...state, networkId };
    case INITIALIZE_METAMASK: {
      const { ethereum } = state;
      // TODO: Move to a setup/env config
      const FYGHTERS_CONTRACT_ADDRESS =
        "0x49de9b5f6c0Dc3e22e9Af986477Cac01dBe82659";
      const provider = new ethers.providers.Web3Provider(ethereum);
      const contract: Fyghters = FyghtersFactory.connect(
        FYGHTERS_CONTRACT_ADDRESS,
        // TODO: Change signer when accounts change
        provider.getSigner()
      );

      return { ...state, contract, provider, ethereum };
    }
    default:
      return state;
  }
};

export const rootReducer = (
  state: FyghtContextInterface = initialState,
  action: Action
): FyghtContextInterface => {
  const { myFyghter, enemies, metamask } = state;
  return {
    myFyghter: myFyghterReducer(myFyghter, action),
    enemies: enemiesReducer(enemies, action),
    metamask: metamaskReducer(metamask, action),
  };
};
