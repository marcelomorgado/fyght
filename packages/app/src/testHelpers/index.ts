import { BigNumber } from "ethers/utils";

export const storeMocks = {
  myFyghter: {
    id: new BigNumber("1"),
    name: "John",
    // TODO: Skin as constant (?)
    skin: "normal_guy",
    xp: new BigNumber("5"),
  },
  enemies: [
    {
      id: new BigNumber("2"),
      name: "Charlie",
      // TODO: Skin as constant (?)
      skin: "naked",
      xp: new BigNumber("1"),
    },
    {
      id: new BigNumber("3"),
      name: "Saul",
      // TODO: Skin as constant (?)
      skin: "ninja",
      xp: new BigNumber("10"),
    },
    {
      id: new BigNumber("4"),
      name: "Julian",
      // TODO: Skin as constant (?)
      skin: "naked",
      xp: new BigNumber("1"),
    },
    {
      id: new BigNumber("5"),
      name: "Brad",
      // TODO: Skin as constant (?)
      skin: "master",
      xp: new BigNumber("89"),
    },
    {
      id: new BigNumber("6"),
      name: "Frank",
      // TODO: Skin as constant (?)
      skin: "no_one",
      xp: new BigNumber("71"),
    },
    {
      id: new BigNumber("7"),
      name: "Brian",
      // TODO: Skin as constant (?)
      skin: "monk",
      xp: new BigNumber("30"),
    },
    {
      id: new BigNumber("8"),
      name: "Fred",
      // TODO: Skin as constant (?)
      skin: "karate_kid",
      xp: new BigNumber("10"),
    },
  ],
};
