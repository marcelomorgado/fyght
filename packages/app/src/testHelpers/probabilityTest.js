const p = (myXp, hisXp) => 1000 / (myXp + hisXp);

for (let myXp = 1; myXp <= 100; ++myXp) {
  for (let hisXp = 1; hisXp <= 100; ++hisXp) {
    console.log(`myXp=${myXp}, hisXp=${hisXp}, p=${p(myXp, hisXp)}`);
  }
}
