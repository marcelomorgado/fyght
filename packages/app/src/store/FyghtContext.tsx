import React, { useReducer, useContext } from "react";
import { createActions } from "./actions";
import { rootReducer, initialState } from "./reducers";

const FyghtContext = React.createContext<any | null>(null);

const asyncer = (dispatch: any, state: any) => (action: any) =>
  typeof action === "function" ? action(dispatch, state) : dispatch(action);

const FyghtProvider = ({ children }: any) => {
  const [state, baseDispatch] = useReducer(rootReducer, initialState);

  const dispatch = React.useCallback(asyncer(baseDispatch, state), []);

  const actions = createActions(dispatch, state);

  return (
    <FyghtContext.Provider value={{ state, ...actions }}>
      {children}
    </FyghtContext.Provider>
  );
};

const useFyghtContext = () => useContext(FyghtContext);

export { FyghtProvider, useFyghtContext };
