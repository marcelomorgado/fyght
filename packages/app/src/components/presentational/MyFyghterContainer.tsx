import React, { useEffect, useState } from "react";
import { Row, Col, Divider } from "antd";
import { MyFyghter } from "./MyFyghter";
import { useFyghtContext } from "../../store";
import { FyghterCreationModal } from "./FyghterCreationModal";

export const MyFyghterContainer: React.FC = () => {
  const [isLoading, setLoading] = useState(true);

  const {
    state: {
      myFyghter,
      metamask: { account },
    },
    loadMyFyghter,
  } = useFyghtContext();

  useEffect(() => {
    const init = async (): Promise<void> => {
      loadMyFyghter();
      setLoading(false);
    };
    init();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [account]);

  const hasFyghter = !isLoading && myFyghter !== null;

  return (
    <>
      <Divider
        orientation="left"
        style={{ color: "#333", fontWeight: "normal" }}
      >
        My Fyghter
      </Divider>
      <Row gutter={[16, 24]}>
        <Col span={24}>
          {isLoading ? (
            <>{`Loading...`}</>
          ) : !hasFyghter ? (
            <FyghterCreationModal />
          ) : (
            <MyFyghter fyghter={myFyghter} />
          )}
        </Col>
      </Row>
    </>
  );
};
