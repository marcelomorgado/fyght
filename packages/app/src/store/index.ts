import { FyghtProvider, useFyghtContext } from "./FyghtContext";
export { FyghtProvider, useFyghtContext };
