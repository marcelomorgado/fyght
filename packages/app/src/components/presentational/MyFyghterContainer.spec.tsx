import * as React from "react";
import { shallow } from "enzyme";
import * as FyghterContext from "../../store";
import { MyFyghterContainer } from "./MyFyghterContainer";
import { storeMocks } from "../../testHelpers";

// TODO: Create a alread loaded test case
describe("MyFyghterContainer", () => {
  test("should render the component", () => {
    const contextValues = { state: { myFyghter: storeMocks.myFyghter } };

    jest
      .spyOn(FyghterContext, "useFyghtContext")
      .mockImplementation(() => contextValues);

    expect(shallow(<MyFyghterContainer />)).toMatchSnapshot();
  });
});
