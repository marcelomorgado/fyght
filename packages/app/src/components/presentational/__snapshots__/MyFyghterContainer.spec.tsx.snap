// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyFyghterContainer should render the component 1`] = `
<Fragment>
  Loading...
</Fragment>
`;
